Add JWT access token support.

jtattermusch opened the pull request:
This adds support for JWT access tokens to a special flavor of ServiceAccountCredential (named JwtServiceAccountCredential). This allows users to use JWT access tokens to authenticate with a service either directly (by using JwtServiceAccountCredential) or through Application Default Credentials stack.

the JwtServiceAccountCredential  works in two modes:
-- if you don't add any scopes to it, it uses JWT access tokens to authenticate the requests.
-- once scopes are added, the behavior falls back to exactly what ServiceAccountCredential does.

If you are using GoogleCredentials, and IsCreateScopedRequired is true (no scopes present), you'll get the JWT access token behavior. After calling CreateScoped(), you'll get the old ServiceAccountCredential behavior. This is consistent to what implementations in other languages are doing (google nodeJS auth library for example).

A few remarks on the changes I've made:
-- a bit of refactoring in ServiceCredential, ServiceAccountCredential and UserCredential to expose ITokenAccess interface directly (which addresses one of the TODOs we've had)
-- the ITokenAccess has changed and now contains method GetAccessTokenForRequestAsync() which accepts auth URI parameter (required for JWT access token support).
-- stripped lots of code from GoogleCredential, thanks to the interface changes I've made.
-- semantics of ServiceAccountCredential is fully preserved so users shouldn't notice any change.

peleyal commented on Src/GoogleApis.Auth.DotNet4.Tests/OAuth2/GoogleCredentialTests.cs:
```diff
@@ -67,7 +70,40 @@ public void FromStream_ServiceAccountCredential()
+
+            // Without adding scopes, the credential should be generating JWT scopes.
+            string accessToken = await (credential as ITokenAccess).GetAccessTokenForRequestAsync(DummyAuthUri);
+            var parts = accessToken.Split(new[] {'.'}, 3);
```
Do we need to add an asset here that parts.length == 3 or something like that...?

jtattermusch replied:
done.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/GoogleCredential.cs:
```diff
@@ -36,10 +32,24 @@ namespace Google.Apis.Auth.OAuth2
     /// See <see cref="GetApplicationDefaultAsync"/> for the credential retrieval logic.
     /// </para>
     /// </summary>
-    public abstract class GoogleCredential : IConfigurableHttpClientInitializer, ITokenAccess
```
Please explain...

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/GoogleCredential.cs:
```diff
@@ -36,10 +32,24 @@ namespace Google.Apis.Auth.OAuth2
+        private readonly IConfigurableHttpClientInitializer clientInitializer;
+
+        /// <summary>
+        /// Creates a new <c>GoogleCredential</c>.
```
You can leave it in one line, (short comment).

I'll add a todo that credential should be some interface and not just object,

jtattermusch replied:
Done (the short comment).

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/JwtServiceAccountCredential.cs:
```diff
@@ -0,0 +1,121 @@
+﻿/*
+Copyright 2013 Google Inc
```
2015

jtattermusch replied:
Done.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/JwtServiceAccountCredential.cs:
```diff
@@ -0,0 +1,121 @@
+    /// See <see cref="GetAccessTokenForRequestAsync"/> for explanation when JWT access token
+    /// is used and when regular OAuth2 token is used.
+    /// <seealso cref="ServiceAccountCredential"/>
+    /// </para>
```
where is the "para" started? You are closing it here, but I don't see where you added it.

jtattermusch replied:
Removed.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/JwtServiceAccountCredential.cs:
```diff
@@ -0,0 +1,121 @@
+        }
+
+        /// <summary>
+        /// JWT access token
```
"." in the end. One line please

jtattermusch replied:
Done.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/JwtServiceAccountCredential.cs:
```diff
@@ -0,0 +1,121 @@
+    /// <seealso cref="ServiceAccountCredential"/>
+    /// </para>
+    /// </summary>
+    public class JwtServiceAccountCredential : ServiceAccountCredential
```
Why can we just extend ServiceAccountCredential to include this "feature" as well?
It might be confusing, (it is to me....)

jtattermusch replied:
I don't have to extends ServiceAccountCredential, if you are fine with changing ServiceAccountCredential's behavior in case there are no scopes assigned (I wasn't sure if service account credential without any scopes can be useful under some weird circumstances).

I also feel just adding functinality to ServiceAccountCredential might be cleaner thing to do here.

jtattermusch replied:
What I meant is I don't have to add the JwtServiceAccountCredential subclass - everything can be implemented directly in ServiceAccountCredential.

jtattermusch replied:
Ok, JwtServiceAccountCredential class is gone. All the code for JWT access tokens is now in ServiceAccountCredential directly.

jtattermusch replied:
I also checked with auth guys that ServiceAccountCredential with no scopes is not something useful (because you can't really use Oauth2 token without a scope for anything). So by changing the pre-existing behavior for case with no scopes we will not be breaking anyone. We will just be adding functionality.

peleyal replied:
AWESOME!
:+1:

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/JwtServiceAccountCredential.cs:
```diff
@@ -0,0 +1,121 @@
+        /// <summary>
+        /// JWT access token
+        /// </summary>
+        internal class JwtAccessToken
```
Why can't you use https://github.com/google/google-api-dotnet-client/blob/master/Src/GoogleApis.Auth/OAuth2/Responses/TokenResponse.cs?

jtattermusch replied:
This is not a Token i received from any token server, so TokenResponse might be a bit misleading. 
I'm fine using it if you think it's alright.

Also, right now I don't technically need to put the access token in a class and I can just use a string (until the moment the caching functionality will be implemented). So if you want a smaller PR, the code can go away.

jtattermusch replied:
I technically wasn't using the JwtAccessToken class so I removed it for now.

peleyal commented on Src/GoogleApis.Auth/OAuth2/UserCredential.cs:
```diff
@@ -91,17 +91,9 @@ public UserCredential(IAuthorizationCodeFlow flow, string userId, TokenResponse
         /// has expired. If successful, it will call <see cref="IAccessMethod.Intercept"/>.
         /// </summary>
-        public async Task InterceptAsync(HttpRequestMessage request, CancellationToken taskCancellationToken)
+        public async Task InterceptAsync(HttpRequestMessage request, CancellationToken cancellationToken)
```
Why do you need this renaming, let's try to keep not necessary changes out of this one.
By the way, I wanted to have "taskCancellationToken", since it might be confusing "token"... is it access token? refresh token? what is cancellation token.. so I just added task as prefix...

jtattermusch replied:
i noticed there's an inconsistency between UserCredential and other credential types. That's fine, I'll remove the change.

jtattermusch replied:
Removed.

peleyal commented on Src/GoogleApis.Auth/OAuth2/ITokenAccess.cs:
```diff
@@ -22,19 +22,24 @@
     /// <see cref="T:Google.Apis.Services.BaseClientService"/> to access the API.
     /// (e.g. gRPC that implemenents the entire HTTP2 stack internally).
     /// </summary>
     public interface ITokenAccess
```
I would be happy to see TokenResponse reminds the same here + we might want to have another interface that inherit from IToeknAccess, IConfigurableHttpClientInitializer and IHttpUnsuccessfulResponseHandler.
Something like ICredential

jtattermusch replied:
I added the ICredential interface with the behavior we agreed on today.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/GoogleCredential.cs:
```diff
@@ -120,107 +130,52 @@ public virtual GoogleCredential CreateScoped(IEnumerable<string> scopes)
-        /// <summary>Wraps <c>ServiceAccountCredential</c> as <c>GoogleCredential</c>.</summary>
-        internal class ServiceAccountGoogleCredential : GoogleCredential
+        /// <summary>Wraps <c>JwtServiceAccountCredential</c> as <c>GoogleCredential</c>.</summary>
+        internal class JwtServiceAccountGoogleCredential : GoogleCredential
```
Why do we need this class, while we deleted the others?
Let's chat. Feel free to put something in my cal before I'm leaving...

jtattermusch replied:
this one overrides the default behavior for IsCreateScopeRequired and CreateScoped. So that makes sense to me.

peleyal commented on Src/GoogleApis.Auth/OAuth2/ITokenAccess.cs:
```diff
@@ -22,19 +22,24 @@
 namespace Google.Apis.Auth.OAuth2
 {
     /// <summary>
```
IMPORTANT!!!!
This interface should be in the Google.Apis.Auth and not in the DotNet4 project.
I didn't notice that before. My bad...

peleyal replied:
Same for https://github.com/google/google-api-dotnet-client/blob/592fef8270cc83f0ef9477069d8da9ba21f40cf5/Src/GoogleApis.Auth.DotNet4/OAuth2/JsonCredentialParameters.cs

Sorry I didn't notice that before.

jtattermusch replied:
Ok, ITokenAccess I already moved to the PCL project.  I also moved JsonCredentialParameter, but I needed to make it public to be able to do that (it was internal before). I am happy to move JsonCredentialParameter back if you agree (I don't see a reason why we should expose this class publicly).

peleyal commented on Src/GoogleApis.Auth.DotNet4.Tests/OAuth2/GoogleCredentialTests.cs:
```diff
@@ -67,7 +70,40 @@ public void FromStream_ServiceAccountCredential()
+            var credential = GoogleCredential.FromStream(stream);
+
+            // Without adding scopes, the credential should be generating JWT scopes.
+            string accessToken = await (credential as ITokenAccess).GetAccessTokenForRequestAsync(DummyAuthUri);
```
Why do you need the conversion? If it's only because you hide that method under the interface I would suggest that you will remove the ITokenAccess from methods when you implement them. I mean that
change:
Task<string> ITokenAccess.GetAccessTokenForRequestAsync(string authUri, CancellationToken cancellationToken)
to:
Task<string> GetAccessTokenForRequestAsync(string authUri, CancellationToken cancellationToken)

I don't remember myself hiding the interface method ever. I'll try to read more about it.
After reading a little bit about it... I'm still not convinced. Why do you prefer to do so?

jtattermusch replied:
I think this is clearer for the user.

Imagine you obtain a GoogleCredential object using GetApplicationDefaultCredential().   Now what you want to do is maybe set scopes to it and maybe pass it to BaseClientService.Initializer or to a gRPC specific auth class.

In none of those case you actually need to call the methods provided IConfigurableHttpClient initializer or ITokenAccess - you just use the fact that the credential object inherits from those interfaces, but you don't wanna call the methods provided by it. On the other hand, the BaseClientService will see the credential as an instance of IConfigurableHttpClient - so it will see the explictly implemented methods right away.

So, I think explicitly inheriting from an interface is what we should do here - it hides the methods the user will never need to call himself from him (he can still call them if he chooses to - but they won't be visible in the common use case), on the other hand, the internal code that will need to call those can call them with no extra effort.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/GoogleCredential.cs:
```diff
@@ -120,107 +130,52 @@ public virtual GoogleCredential CreateScoped(IEnumerable<string> scopes)
         internal static GoogleCredential FromCredential(ComputeCredential credential)
         {
-            return new ComputeGoogleCredential(credential);
+            return new GoogleCredential(credential, credential, credential);
```
Bahhhhhhh.
There should be one interface ICredential
That's one of my biggest concerns in this cl.

jtattermusch replied:
I can try to introduce ICredential : IConfigurableHttpInitializer, ITokenAccess

i think that's what you are suggesting anyway.

peleyal commented:
3 things to consider:

1. An interface ICredential
return new GoogleCredential(credential, credential, credential);
protected GoogleCredential(object credential, ITokenAccess tokenAccess, IConfigurableHttpClientInitializer clientInitializer)

Bahhh....

The credential is ItoeknAccess and IConfigurableHttpClientInitializer!
We should introduce a new interface that wraps the 3 together (we might want to include UnsuccessfulResponseHandler as well)

2. JwtServiceAccount should be merged to ServiceAccount
Not sure why we need Jwt specific class.
Why do we need to declare a new JwtAccessToken class? We should use the TokenResponse.

3. Some classes should be moved to Google.Apis.Auth (not dotnet4)
https://github.com/google/google-api-dotnet-client/blob/592fef8270cc83f0ef9477069d8da9ba21f40cf5/Src/GoogleApis.Auth.DotNet4/OAuth2/ITokenAccess.cs
and
https://github.com/google/google-api-dotnet-client/blob/592fef8270cc83f0ef9477069d8da9ba21f40cf5/Src/GoogleApis.Auth.DotNet4/OAuth2/JsonCredentialParameters.cs
They are not for a specific platform. They should be in the PCL project.


Let's chat ASAP, before I'm leaving.

peleyal commented on Src/GoogleApis.Auth/OAuth2/UserCredential.cs:
```diff
@@ -91,17 +91,9 @@ public UserCredential(IAuthorizationCodeFlow flow, string userId, TokenResponse
-                }
-            }
-
+            var accessToken = await GetAccessTokenForRequestAsync(request.RequestUri.ToString(), cancellationToken);
```
ConfigureAwait

jtattermusch replied:
Done.

peleyal commented on Src/GoogleApis.Auth/OAuth2/UserCredential.cs:
```diff
@@ -133,6 +125,27 @@ public void Initialize(ConfigurableHttpClient httpClient)
+        /// Gets an access token to authorize a request. If the existing token has expired, try to refresh it first.
+        /// <seealso cref="ITokenAccess.GetAccessTokenForRequestAsync"/>
+        /// </summary>
+        public virtual async Task<string> GetAccessTokenForRequestAsync(string authUri = null, CancellationToken cancellationToken = default(CancellationToken))
```
Remove doc here. We inherit it form the interface

jtattermusch replied:
Done.

jtattermusch commented:
Ok, so:
1.  I introduced the ICredential interface.
2. Moved the classes you wanted (see my comment for JsonCredentialParameters)
3. Got rid of JwtServiceAccountCredential.

I believe this addresses all the comments you've had.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/GoogleCredential.cs:
```diff
@@ -36,10 +32,18 @@ namespace Google.Apis.Auth.OAuth2
+        private readonly ICredential credential;
+
+        /// <summary>Creates a new <c>GoogleCredential</c>.</summary>
+        internal GoogleCredential(ICredential credential)
```
Wooohooooo!
Looks much much better now :) :+1:

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/GoogleCredential.cs:
```diff
@@ -36,10 +32,18 @@ namespace Google.Apis.Auth.OAuth2
-    public abstract class GoogleCredential : IConfigurableHttpClientInitializer, ITokenAccess
+    public class GoogleCredential : IConfigurableHttpClientInitializer, ITokenAccess
     {
         private static DefaultCredentialProvider defaultCredentialProvider = new DefaultCredentialProvider();
```
Consider adding comments for this field and for ICredential below as well.
Yeah I know it's private, but why not... :)

jtattermusch replied:
Done.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/GoogleCredential.cs:
```diff
@@ -120,107 +124,40 @@ public virtual GoogleCredential CreateScoped(IEnumerable<string> scopes)
-        }
-
         /// <summary>Wraps <c>ServiceAccountCredential</c> as <c>GoogleCredential</c>.</summary>
         internal class ServiceAccountGoogleCredential : GoogleCredential
```
I would try to avoid this class (if you can)
If you see no other way, I would try to emphasize why we do need it, while we don't have a specific GoogleCredential class for compute or user credential.

jtattermusch replied:
I spent some time experimenting with replacing this with passing delegates providing special handling for IsCreateScopedRequired and Create Scoped to the constructor and the result didn't look good. It resulted in more code being added and it was pretty unclear because I was basically trying to reinvent inheritance.
Leaving ServiceAccountGoogleCredential class, but added a comment clarify why we are subclassing.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/ServiceAccountCredential.cs:
```diff
@@ -1,5 +1,5 @@
 ﻿/*
-Copyright 2013 Google Inc
+Copyright 2015 Google Inc
```
You don't need to change the year in case you are changing a file.
You only add 2015 to new files :)

jtattermusch replied:
ok :-)

jtattermusch replied:
Done.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/GoogleCredential.cs:
```diff
@@ -120,107 +124,40 @@ public virtual GoogleCredential CreateScoped(IEnumerable<string> scopes)
             public ServiceAccountGoogleCredential(ServiceAccountCredential credential)
+                : base(credential)
             {
                 this.credential = credential;
```
Please delete. You don't need that, you already inherit credential from GoogleCredential.
If you need to use credential as ServiceAccountCredentail just cast it.

jtattermusch replied:
Done.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/GoogleCredential.cs:
```diff
@@ -229,7 +166,7 @@ public ServiceAccountGoogleCredential(ServiceAccountCredential credential)
             public override bool IsCreateScopedRequired
             {
-                get { return (credential.Scopes == null || credential.Scopes.Count() == 0); }
+                get { return !credential.HasScopes; }
```
!(credential as ServiceAccountCredentail).HasScopes

jtattermusch replied:
Done.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/ServiceAccountCredential.cs:
```diff
@@ -41,6 +41,14 @@ namespace Google.Apis.Auth.OAuth2
     /// Take a look in https://developers.google.com/accounts/docs/OAuth2ServiceAccount for more details.
     /// </para>
+    /// <para>
+    /// Service account credential also supports JSON Web Token access token scenario.
```
You can also add a prefix: From version 1.9.3, ...
So: "From version 1.9.3, service account credential also....

jtattermusch replied:
done.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/ServiceAccountCredential.cs:
```diff
@@ -41,6 +41,14 @@ namespace Google.Apis.Auth.OAuth2
+    /// <para>
+    /// Service account credential also supports JSON Web Token access token scenario.
+    /// In this scenario, instead of sending a JWT to a token server and exchanging it for 
+    /// an access token, a slightly different JWT containing the request URI is constructed
```
"a slightly different JWT" elaborate or rephrase.
Different how?

jtattermusch replied:
Rephrased.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/ServiceAccountCredential.cs:
```diff
@@ -180,9 +178,75 @@ public ServiceAccountCredential(Initializer initializer) : base(initializer)
+            return assertion.ToString();
+        }
+
+        // TODO: fix stale link
```
mmm not sure what do you mean here. If it's something for you to fix later on add your name.

jtattermusch replied:
Filed a separate issue to update the out of date links throughout the code. Removed todos from this PR.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/ServiceAccountCredential.cs:
```diff
@@ -180,9 +178,75 @@ public ServiceAccountCredential(Initializer initializer) : base(initializer)
+        /// Signs JWT token using the private key and returns the serialized assertion.
+        /// </summary>
+        /// <param name="payload">the JWT payload to sign.</param>
+        private string CreateAssertionFromPayload(JsonWebSignature.Payload payload)
```
It was hard for me to follow the small changes here. Make sure you test it, and I would also recommend making sure that our G+ sample (https://github.com/google/google-api-dotnet-client-samples/tree/master/Plus.ServiceAccount) is working with the new API, since you are changing the service account functionality + refactoring stuff.. Just to make sure it works as it used to be :)

jtattermusch replied:
Will do.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/ServiceAccountCredential.cs:
```diff
@@ -180,9 +178,75 @@ public ServiceAccountCredential(Initializer initializer) : base(initializer)
+            var payload = new JsonWebSignature.Payload()
+            {
+                Issuer = Id,
+                Subject = Id,
```
Both subject and Issuer are Id?
Just making sure (in the other case we used Subject = User)...

You know better than I, just want to make sure that you didn't miss that one

jtattermusch replied:
Yes, that's up to the spec.

peleyal commented on Src/GoogleApis.Auth/OAuth2/ICredential.cs:
```diff
@@ -0,0 +1,25 @@
+
+namespace Google.Apis.Auth.OAuth2
+{
+    /// <summary>Functionality common to all credential types.</summary>
```
It would be nice to elaborate more.

ICredential is the main interface to represent credential in the client library. Service account, User account and Compute credential inherits from this interface to provide access token functionality. In addition this interface inherits form configurable http client initializer (link...) to be able to hook to http requests. More details are available in the specific implementations.

jtattermusch replied:
Added your comment and changed the wording a bit. Thanks!

peleyal commented on Src/GoogleApis.Auth/OAuth2/ITokenAccess.cs:
```diff
@@ -22,19 +22,24 @@
+        /// migth just ignore it.
+        /// </summary>
+        /// <param name="authUri">The URI the returned token will grant access to.</param>
+        /// <param name="cancellationToken">The cancellation token.</param>
```
Don't you feel that it's confusing, is it a task token? or is it an access token...?
That's why I added task prefix before...

Just saying, it's up to you to rename it or leave it like that

jtattermusch replied:
To me it's not confusing at all. CancellationToken is a widely known concept so cancellationToken variable seems pretty clear to me. Actually, it's used throughout the existing code in numerous places. Leaving as is.

peleyal commented:
Change looks much much better now, after creating ICredential.
I would be happy also to see how can we delete ServiceAccountGoogleCredential if it's possible.

Gus, most of the change looks good to me, worth to add tests (when possible).
Your opinion is valuable, I won't be able to have another review, so do your best :) :+1:

jtattermusch commented:
Okay, I've addressed all the comments.
I wasn't able to remove the ServiceAccountGoogleClass in a way I would be happy with, so I just added explanatory comment why we need it (see one of the previous comments for further explanation).

gguuss commented on Src/GoogleApis.Auth/OAuth2/ServiceCredential.cs:
```diff
@@ -197,6 +187,29 @@ public void Initialize(ConfigurableHttpClient httpClient)
+                }
+                Logger.Info("New access token was received successfully");
+            }
+            return Token.AccessToken;
```
Above you have a null check for Token, can Token be null here too or will the token just be null upon initialization and then on this check it gets created?

jtattermusch replied:
This is actually preexisting code, I just refactored the call sites a bit. I think it's fine without the second null check because once RequestAccessTokenAsync  succeeds, it will always set the Token. If it fails, this method throws exception before it reaches line 208.

gguuss commented on Src/GoogleApis.Auth/OAuth2/UserCredential.cs:
```diff
@@ -133,6 +124,23 @@ public void Initialize(ConfigurableHttpClient httpClient)
+
+        public virtual async Task<string> GetAccessTokenForRequestAsync(string authUri = null, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            if (Token.IsExpired(flow.Clock))
```
Above you protected the check on IsExpired with a null check, maybe add one here too if it can be null?

jtattermusch replied:
I think for user credential (unlike service credential) the Token is never null. It gets assigned in the constructor. 
Again, this is preexisting code.

jtattermusch commented:
I received an LGTM offline. Fine to merge?

gguuss commented:
LGTM
